src/compiler/emitter.ts

```diff
@@ -3826,78 +3883,158 @@ module ts {
+                        if (nodeIsDecorated(parameter)) {
+                            writeComma = emitDecoratorArray(parameter.decorators, writeComma, `__param(${parameterIndex}, `, ")");
+                        }
+                        ++parameterIndex;
                     }
+                }
+                return writeComma;
+            }
 
-                    // Emit the decorators for a parameter. Given the following:
-                    //
-                    //   class C {
-                    //     constructor(@dec p) { }
-                    //     method(@dec p) { }
-                    //     set accessor(@dec value) { }
-                    //   }
-                    //
-                    // The emit for a constructor is:
-                    //
-                    //   __decorate([dec], C, void 0, 0);
-                    //
-                    // The emit for a parameter is:
-                    //
-                    //   __decorate([dec], C.prototype, "method", 0);
-                    //
-                    // The emit for an accessor is:
-                    //
-                    //   __decorate([dec], C.prototype, "accessor", 0);
-                    //
+            function emitDecoratorArray(decorators: NodeArray<Decorator>, writeComma: boolean, prefix?: string, postfix?: string): boolean {
+                if (decorators) {
+                    let decoratorCount = decorators ? decorators.length : 0;
+                    for (let i = 0; i < decoratorCount; i++) {
+                        if (writeComma) {
+                            write(",");
+                        }
+                        writeLine();
 
-                    writeLine();
-                    emitStart(parameter);
-                    emitDecorateStart(parameter.decorators);
-                    emitStart(parameter.name);
+                        let decorator = decorators[i];
+                        emitStart(decorator);
+                        write(prefix);
+                        emit(decorator.expression);
+                        write(postfix);
+                        emitEnd(decorator);
+                        writeComma = true;
+                    }
+                }
+                return writeComma;
+            }
 
-                    if (member.kind === SyntaxKind.Constructor) {
-                        emitDeclarationName(node);
-                        write(", void 0");
+            function shouldEmitTypeMetadata(node: Declaration): boolean {
+                if (!compilerOptions.emitDecoratorMetadata) {
+                    return false;
+                }
+
+                switch (node.kind) {
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.SetAccessor:
+                    case SyntaxKind.PropertyDeclaration:
+                        return true;

```

Comment:
This is only used by `emitSerializedTypeMetadata`, which is only called for a decorator. Adding the check would add unnecessary overhead.